specs/lightclient/beacon-chain.md

```diff
@@ -166,6 +242,92 @@ def get_sync_committee(state: BeaconState, epoch: Epoch) -> SyncCommittee:
     return SyncCommittee(pubkeys=pubkeys, pubkey_aggregates=aggregates)
 ```
 
+#### `get_base_reward`
+
+*Note*: The function `get_base_reward` is modified with the removal of `BASE_REWARDS_PER_EPOCH`.
+
+```python
+def get_base_reward(state: BeaconState, index: ValidatorIndex) -> Gwei:
+    total_balance = get_total_active_balance(state)
+    effective_balance = state.validators[index].effective_balance
+    return Gwei(effective_balance * BASE_REWARD_FACTOR // integer_squareroot(total_balance))
+```
+
+#### `get_unslashed_participating_indices`
+
+```python
+def get_unslashed_participating_indices(state: BeaconState, flag: uint8, epoch: Epoch) -> Set[ValidatorIndex]:
+    assert epoch in (get_previous_epoch(state), get_current_epoch(state))
+    if epoch == get_current_epoch(state):
+        epoch_participation = state.current_epoch_participation
+    else:
+        epoch_participation = state.previous_epoch_participation
+    participating_indices = [
+        index for index in get_active_validator_indices(state, epoch)
+        if epoch_participation[index][flag]
+    ]
+    return set(filter(lambda index: not state.validators[index].slashed, participating_indices))
+```
+
+#### `get_flag_deltas`
+
+```python
+def get_flag_deltas(state: BeaconState, flag: uint8, numerator: uint64) -> Tuple[Sequence[Gwei], Sequence[Gwei]]:
+    """
+    Computes the rewards and penalties associated with a particular duty, by scanning through the participation
+    flags to determine who participated and who did not and assigning them the appropriate rewards and penalties.
+    """
+    rewards = [Gwei(0)] * len(state.validators)
+    penalties = [Gwei(0)] * len(state.validators)
+
+    unslashed_participating_indices = get_unslashed_participating_indices(state, flag, get_previous_epoch(state))
+    increment = EFFECTIVE_BALANCE_INCREMENT  # Factored out from balances to avoid uint64 overflow
+    unslashed_participating_increments = get_total_balance(state, unslashed_participating_indices) // increment
+    active_increments = get_total_active_balance(state) // increment
```

Comment:
`unslashed_participating_increments` refers to previous epoch
`active_increments` refers to current epoch

is that ok?

Reply:
worth noting, this has the same semantics as what is on mainnet.

and given that there is a constant difference of one epoch here, we can be safe knowing each set of balances are referring to the same "dynasties" (cf. Casper FFG paper, contiguous sets of validators) because the activation/exit delay is greater than one epoch. we filter out slashed validators in the numerator as well so we don't have to worry about slashings in this epoch affecting the relevant set of validators either.